src/sender.js

```diff
@@ -431,6 +596,115 @@ function startNewRow(sender) {
     sender.hasColumns = false;
 }
 
+function createRequestOptions(sender, data) {
```

Comment:
We seem to use the default agent so that the HTTP connections will be pooled:
https://nodejs.org/api/http.html#httpglobalagent
https://nodejs.org/api/http.html#class-httpagent

We should document that the connections are pooled and, also, double-check the default idle timeout in the agent, so that they're not greater than the default server-side ones. The default agent doesn't limit the number of connections which isn't great - we limit them with 256 on the server-side.

One more thing to consider would be to use a custom agent and also allow users to provide their own agent in config options - the downside will be having to release the idle connections when the last active sender is closed.

Reply:
Reading the doc, and it seems to me that:
 - connections are not pooled by default, `keepAlive` defaults to `false` in the agent,
 - but as you said, the number of active connections are not limited, `maxSockets` defaults to `Infinity`.
 
I also do not see an idle timeout option, so it is unlikely that it could be set.
As I understand it based on the doc, if the connection is pooled, it stays in the pool until the server closes it.
When it is closed, it is removed from the pool.

Probably best is to give the option to provide a custom agent, so they can modify any of the options as they wish.
By default we will set `maxSockets` to 256 via our own custom agent.

Re "the downside will be having to release the idle connections when the last active sender is closed", this only matters if they set keepAlive=true, and we could just let the server to close them after the idle timeout anyway.
I imagine we are talking about the situation when their client is 'stateless', what i mean is they create a new sender for every message.
In this case it can easily happen that no sender exists, but in the next moment a new message is processed, and a sender is created again. We could end up with emptying the pool for no reason, especially when messages arrive in bursts.

Reply:
added a custom agent option

Reply:
> * connections are not pooled by default, `keepAlive` defaults to `false` in the agent,

It's more complicated:
![Screenshot from 2024-04-24 09-59-51](https://github.com/questdb/nodejs-questdb-client/assets/37772591/60a79657-ed3e-41d0-a1c9-79674988d7e9)

Reply:
> I also do not see an idle timeout option, so it is unlikely that it could be set.

There are `keepAliveMsecs` and `timeout` options in the agent. While the docs aren't crystal clear, the latter is likely what we need.

Reply:
> As I understand it based on the doc, if the connection is pooled, it stays in the pool until the server closes it.
When it is closed, it is removed from the pool.

No, the client also checks connection timeouts, closes them and removes them from the pool.

Reply:
> > I also do not see an idle timeout option, so it is unlikely that it could be set.
> 
> There are `keepAliveMsecs` and `timeout` options in the agent. While the docs aren't crystal clear, the latter is likely what we need.

`keepAliveMsecs` is definitely not the idle timeout.
From the docs `timeout` looked like a default for the http request timeout to me, but looking at the code in this node PR, it is the idle timeout: https://github.com/nodejs/node/pull/43522/files#diff-82c531677b9ab4e260025e0397fb0ac4d4316459fa5d16346324e39584cfce96

Reply:
I will change the custom agent to default to keepAlive=true regardless of node version, and will set a 1 minute default idle timeout (we have 5 minutes default on server side).
I think this should work for the majority of users.
If not, they can pass in their own agent with their own config.

Reply:
pushed the change

Reply:
There is one thing we should do with the default (not user-provided) custom agent: namely we should introduce a counter of active sender instances and check it when a sender is closed. If the counter goes to zero, we should call `agent.destroy()` to avoid resource leak in situations when senders are actively used and then there is a large inactivity period in the application.
https://nodejs.org/api/http.html#agentdestroy

We do a similar thing in the Go client:
https://github.com/questdb/go-questdb-client/blob/2190485745fb270465fabf6af364bcb7dee3b61a/http_sender.go#L316

Reply:
> There is one thing we should do with the default (not user-provided) custom agent: namely we should introduce a counter of active sender instances and check it when a sender is closed. If the counter goes to zero, we should call `agent.destroy()` to avoid resource leak in situations when senders are actively used and then there is a large inactivity period in the application. https://nodejs.org/api/http.html#agentdestroy
> 
> We do a similar thing in the Go client: https://github.com/questdb/go-questdb-client/blob/2190485745fb270465fabf6af364bcb7dee3b61a/http_sender.go#L316

Is this really necessary? Sockets have their idle timeout, and will be closed eventually.
Also, it might not be beneficial to destroy the agent immediately when there are no sender objects.
When messages arrive in bursts few seconds apart, for example, we would close all connections after each burst, then reconnect few seconds later. Or if sometimes there is a period when no messages are processed, we will immediately close all connections, but the next message could come shortly after.
The idle timeout caters for these situations.

However, you are probably right in that we should provide a way to destroy the agent.
Maybe a static `Sender.destroy()` method?

Reply:
"Maybe a static Sender.destroy() method?" -> This does not look too nice, when it is used.
I mean you need to close all senders, and then also call the static destroy() at the end, which is easy to forget.
I am changing my mind, and will add the static counter as you suggested.
The API will be cleaner, and I prefer that.

Reply:
done

Reply:
> Is this really necessary? Sockets have their idle timeout, and will be closed eventually.

The thing is that sockets will be closed only on the next sender creation and usage and that's up to the user application.

Reply:
I thought they would register a callback for the 'closed' event, and when the server disconnects they could act on it. Even the doc says that it is removed if closed by the server:
"When a connection is closed by the client or the server, it is removed from the pool. Any unused sockets in the pool will be unrefed so as not to keep the Node.js process running when there are no outstanding requests. (see [socket.unref()](https://nodejs.org/api/net.html#socketunref))."

But it also says that:
"It is good practice, to [destroy()](https://nodejs.org/api/http.html#agentdestroy) an Agent instance when it is no longer in use, because unused sockets consume OS resources."
So we should definitely call destroy() one way or another, and calling it after the last sender instance is closed seems to be the best time to do it.

Reply:
> Even the doc says that it is removed if closed by the server

Yes, it looks like that's the case:
https://github.com/nodejs/node/blob/708bffa9992d65fbe063c6a2f30892b8b5c0af57/lib/_http_agent.js#L395-L407